File: products/sdlc-cc/platform/packages/insights-core/go/exec/adapter.go
package exec

import (
	"context"
	"errors"

	"github.com/sdlc-ai/platform/packages/insights-core/types"
)

type Capabilities struct {
	Idempotent       bool
	RequiresApproval bool
}

type Preview struct {
	Summary string         `json:"summary"`
	Payload map[string]any `json:"payload"`
}

// Adapter is the contract implemented by every exec target (Jira, Slack, …).
// Full impls arrive in T-108+; T-107 implements the router that drives this.
type Adapter interface {
	Name() string
	Capabilities() Capabilities
	Validate(params map[string]any) error
	DryRun(ctx context.Context, ins types.Insight, params map[string]any) (Preview, error)
	Execute(ctx context.Context, ins types.Insight, params map[string]any) (types.Receipt, error)
}

// Registry holds the active adapters, keyed by Name().
type Registry struct {
	adapters map[string]Adapter
}

func NewRegistry() *Registry { return &Registry{adapters: map[string]Adapter{}} }

func (r *Registry) Register(a Adapter) error {
	if a == nil {
		return errors.New("adapter nil")
	}
	if _, exists := r.adapters[a.Name()]; exists {
		return errors.New("adapter already registered: " + a.Name())
	}
	r.adapters[a.Name()] = a
	return nil
}

func (r *Registry) Get(name string) (Adapter, bool) {
	a, ok := r.adapters[name]
	return a, ok
}

func (r *Registry) Names() []string {
	out := make([]string, 0, len(r.adapters))
	for k := range r.adapters {
		out = append(out, k)
	}
	return out
}

File: products/sdlc-cc/platform/packages/insights-core/ts/src/adapter.ts
import type { Insight, Receipt } from "./types.js";

export interface Capabilities {
  idempotent: boolean;
  requires_approval: boolean;
}

export interface Preview {
  summary: string;
  payload: Record<string, unknown>;
}

export interface Adapter {
  name(): string;
  capabilities(): Capabilities;
  validate(params: Record<string, unknown>): void;
  dryRun(ins: Insight, params: Record<string, unknown>): Promise<Preview>;
  execute(ins: Insight, params: Record<string, unknown>): Promise<Receipt>;
}

export class Registry {
  private readonly adapters = new Map<string, Adapter>();

  register(a: Adapter): void {
    if (this.adapters.has(a.name())) {
      throw new Error(`adapter already registered: ${a.name()}`);
    }
    this.adapters.set(a.name(), a);
  }

  get(name: string): Adapter | undefined {
    return this.adapters.get(name);
  }

  names(): string[] {
    return [...this.adapters.keys()];
  }
}
